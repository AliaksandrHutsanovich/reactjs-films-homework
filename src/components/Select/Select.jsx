import React, { memo, useMemo } from 'react';
import { Select as AntSelect } from 'antd';
import { DownOutlined } from '@ant-design/icons';
import PropTypes from 'prop-types';

import '!style-loader!css-loader!sass-loader!antd/dist/antd.css'; //eslint-disable-line
import styles from './Select.scss';

const { Option } = AntSelect;

const defaultArrowIcon = <DownOutlined className={styles.arrowIcon} />;

export const defaultRenderOption = (option) => (
  <Option
    key={option.id}
    className={styles.option}
    value={option.id}
  >
    {option.name}
  </Option>
);

const Select = ({
  placeholder,
  bordered,
  allowClear,
  arrowIcon,
  onSelect,
  options,
  renderOption,
  disabled,
}) => (
  <AntSelect
    placeholder={placeholder}
    bordered={bordered}
    allowClear={allowClear}
    suffixIcon={arrowIcon}
    onSelect={onSelect}
    className={styles.select}
    dropdownClassName={styles.dropDown}
    disabled={disabled}
    onChange={onSelect}
    virtual={false}
  >
    {
      useMemo(
        () => (options.map((option) => renderOption(option))),
        [options, renderOption],
      )
    }
  </AntSelect>
);

Select.propTypes = {
  placeholder: PropTypes.string.isRequired,
  bordered: PropTypes.bool,
  allowClear: PropTypes.bool,
  arrowIcon: PropTypes.node,
  onSelect: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
      ]).isRequired,
    }),
  ).isRequired,
  renderOption: PropTypes.func,
  disabled: PropTypes.bool,
};

Select.defaultProps = {
  bordered: false,
  allowClear: false,
  arrowIcon: defaultArrowIcon,
  renderOption: defaultRenderOption,
  disabled: false,
};

export default memo(Select);
