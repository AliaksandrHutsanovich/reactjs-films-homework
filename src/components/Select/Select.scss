@font-face{
  font-family: 'HelveticaNeue';
  src: url('../../assets/fonts/HelveticaNeue.ttf') format('truetype');
}

.select {
  width: fit-content;
  color: #1890ff;
  font: 14px HelveticaNeue;

  span {
    opacity: 1;
  }
}

.arrowIcon {
  color: #1890ff;
  font-size: 10px;
}

.option {
  color: #1890ff;
  width: fit-content;
}

.dropDown {
  width: auto !important;

  > div > div:nth-child(2) > div > div {
    position: static !important;
  }
}


