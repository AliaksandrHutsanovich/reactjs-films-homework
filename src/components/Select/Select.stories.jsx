import React from 'react';
import { action } from '@storybook/addon-actions';
import Select from './Select';

export default {
  component: Select,
  title: 'components/Select',
};

const options = [
  {
    id: 1,
    name: 'Action',
  }, {
    id: 2,
    name: 'Adventure',
  }, {
    id: 3,
    name: 'Fantasy',
  }, {
    id: 4,
    name: 'Sci-fi',
  }, {
    id: 5,
    name: 'Action',
  }, {
    id: 6,
    name: 'Adventure',
  }, {
    id: 7,
    name: 'Fantasy',
  }, {
    id: 8,
    name: 'Sci-fi',
  }, {
    id: 9,
    name: 'Action',
  }, {
    id: 10,
    name: 'Adventure',
  }, {
    id: 11,
    name: 'Fantasy',
  }, {
    id: 12,
    name: 'Sci-fi',
  },
];

export const unborderedSelect = () => (
  <Select
    placeholder="Genre"
    allowClear
    onSelect={action('the option in unbordered select was selected')}
    options={options}
  />
);

export const disabledSelect = () => (
  <Select
    placeholder="Genre"
    allowClear
    onSelect={action('the option in unbordered select was selected')}
    options={options}
    disabled
  />
);
