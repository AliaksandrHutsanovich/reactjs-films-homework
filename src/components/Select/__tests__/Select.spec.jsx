import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Select, { defaultRenderOption } from '../Select';

const options = [
  {
    id: 1,
    name: 'Action',
  }, {
    id: 2,
    name: 'Adventure',
  }, {
    id: 3,
    name: 'Fantasy',
  }, {
    id: 4,
    name: 'Sci-fi',
  }, {
    id: 5,
    name: 'Action',
  }, {
    id: 6,
    name: 'Adventure',
  }, {
    id: 7,
    name: 'Fantasy',
  }, {
    id: 8,
    name: 'Sci-fi',
  }, {
    id: 9,
    name: 'Action',
  }, {
    id: 10,
    name: 'Adventure',
  }, {
    id: 11,
    name: 'Fantasy',
  }, {
    id: 12,
    name: 'Sci-fi',
  },
];

describe('<Select />', () => {
  const renderer = new ShallowRenderer();

  it('unbordered select should render properly', () => {
    renderer.render(
      <Select
        placeholder="Genre"
        allowClear
        onSelect={() => {}}
        options={options}
        renderOption={defaultRenderOption}
      />,
    );
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('disabled select should be rendered properly', () => {
    renderer.render(
      <Select
        placeholder="Genre"
        allowClear
        onSelect={() => {}}
        options={options}
        disabled
        renderOption={defaultRenderOption}
      />,
    );
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
