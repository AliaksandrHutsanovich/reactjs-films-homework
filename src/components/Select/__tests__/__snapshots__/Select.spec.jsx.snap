// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> disabled select should be rendered properly 1`] = `
<Select
  allowClear={true}
  bordered={true}
  choiceTransitionName="zoom"
  className="select"
  disabled={true}
  dropdownClassName="dropDown"
  onChange={[Function]}
  onSelect={[Function]}
  placeholder="Genre"
  transitionName="slide-up"
  virtual={false}
>
  <Option
    className="option"
    value={1}
  >
    Action
  </Option>
  <Option
    className="option"
    value={2}
  >
    Adventure
  </Option>
  <Option
    className="option"
    value={3}
  >
    Fantasy
  </Option>
  <Option
    className="option"
    value={4}
  >
    Sci-fi
  </Option>
  <Option
    className="option"
    value={5}
  >
    Action
  </Option>
  <Option
    className="option"
    value={6}
  >
    Adventure
  </Option>
  <Option
    className="option"
    value={7}
  >
    Fantasy
  </Option>
  <Option
    className="option"
    value={8}
  >
    Sci-fi
  </Option>
  <Option
    className="option"
    value={9}
  >
    Action
  </Option>
  <Option
    className="option"
    value={10}
  >
    Adventure
  </Option>
  <Option
    className="option"
    value={11}
  >
    Fantasy
  </Option>
  <Option
    className="option"
    value={12}
  >
    Sci-fi
  </Option>
</Select>
`;

exports[`<Select /> unbordered select should render properly 1`] = `
<Select
  allowClear={true}
  bordered={true}
  choiceTransitionName="zoom"
  className="select"
  dropdownClassName="dropDown"
  onChange={[Function]}
  onSelect={[Function]}
  placeholder="Genre"
  transitionName="slide-up"
  virtual={false}
>
  <Option
    className="option"
    value={1}
  >
    Action
  </Option>
  <Option
    className="option"
    value={2}
  >
    Adventure
  </Option>
  <Option
    className="option"
    value={3}
  >
    Fantasy
  </Option>
  <Option
    className="option"
    value={4}
  >
    Sci-fi
  </Option>
  <Option
    className="option"
    value={5}
  >
    Action
  </Option>
  <Option
    className="option"
    value={6}
  >
    Adventure
  </Option>
  <Option
    className="option"
    value={7}
  >
    Fantasy
  </Option>
  <Option
    className="option"
    value={8}
  >
    Sci-fi
  </Option>
  <Option
    className="option"
    value={9}
  >
    Action
  </Option>
  <Option
    className="option"
    value={10}
  >
    Adventure
  </Option>
  <Option
    className="option"
    value={11}
  >
    Fantasy
  </Option>
  <Option
    className="option"
    value={12}
  >
    Sci-fi
  </Option>
</Select>
`;
